src/Chapter3.hs

```diff
@@ -903,13 +986,38 @@ We will call such a typeclass "Append". You can find its definition below.
 Implement instances of "Append" for the following types:
 
   ✧ The "Gold" newtype where append is the addition
-  ✧ "List" where append is list concatenation
+  ✧ "List" where append is list concatenation(assuming not [a])
   ✧ *(Challenge): "Maybe" where append is appending of values inside "Just" constructors
 
 -}
 class Append a where
     append :: a -> a -> a
 
+newtype Gold = Gold Int
+getGold :: Gold -> Int
+getGold (Gold x) = x
+
+data List a
+    = Empty
+    | Cons a (List a)
+
+instance Append Gold where
+    append :: Gold -> Gold -> Gold
+    append x y = Gold (x_i + y_i)
+      where x_i = getGold x
+            y_i = getGold y
+
+instance Append a => Append (List a) where
+  append :: List a -> List a -> List a
+  append Empty ys = ys
+  append (Cons x xs) ys =  Cons x (append xs ys)
+
+-- Why can't I "append :: Maybe a -> Maybe a -> a", where (Nothing, Nothing) gives error("nullptr")?
+instance Append a => Append (Maybe a) where
+  append :: Maybe a -> Maybe a -> Maybe a
```

Comment:
The type signature of the `Append` typeclass forces to have the resulting type the same as the input type.

```haskell
append :: a -> a -> a
```

So, when you need to return `Maybe`, it looks reasonable `Nothing` on `(Nothing, Nothing)`. Generally, errors are discouraged, and it's usually better to avoid them when possible.

If you want to learn more about the `Append` typeclass, you can look at the `Semigroup` typeclass from the standard library, which has the same meaning:

* http://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Semigroup.html#t:Semigroup